### PR TITLE
[SPARK-22829] Add new built-in function date_trunc()

### PR DESCRIPTION
**Pull request by youngbink**
## What changes were proposed in this pull request?

Adding date_trunc() as a built-in function.
`date_trunc` is common in other databases, but Spark or Hive does not have support for this. `date_trunc` is commonly used by data scientists and business intelligence application such as Superset (https://github.com/apache/incubator-superset).
We do have `trunc` but this only works with 'MONTH' and 'YEAR' level on the DateType input.

date_trunc() in other databases:
AWS Redshift: http://docs.aws.amazon.com/redshift/latest/dg/r_DATE_TRUNC.html
PostgreSQL: https://www.postgresql.org/docs/9.1/static/functions-datetime.html
Presto: https://prestodb.io/docs/current/functions/datetime.html

## How was this patch tested?

Unit tests

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by youngbink** [comment]
@gatorsmile @cloud-fan

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
we need to create a JIRA ticket

**Comment by HyukjinKwon** [comment]
@cloud-fan and @youngbink how about reviving https://github.com/apache/spark/pull/14788 with a configuration to control this?

AWS Redshift seems having `TRUNC` which just converts a timestamp to a date whereas we have Spark's `trunc` where supports date formats. This is not quite equivalent. I think Spark's `trunc` is more like Redshift's `DATE_TRUNC`.

PostgreSQL does not have `trunc` but has `date_trunc` where we can specify the format and returns a timestamp always.

Presto also looks not having a duplicated functionality.

I think we can simply introduce an alias for `trunc` after resolving https://github.com/apache/spark/pull/14788.

Did I maybe miss something?

**Comment by youngbink** [comment]
@HyukjinKwon  Just took a look at this PR #14788. 

My point of mentioning those databases was just to give examples of the function that Spark doesn't support but other databases commonly do. (They all have this `date_trunc` which takes `timestamp` and output `timestamp`)
As you said, we could extend `trunc` and simply create an alias `date_trunc`, but it's actually not as simple. For e.g, PR #14788 won't be able to handle the following command collectly on PySpark:
```
df = spark.createDataFrame([('1997-02-28 05:02:11',)], ['d'])
df.select(functions.trunc(df.d, 'year').alias('year')).collect()  
df.select(functions.trunc(df.d, 'SS').alias('SS')).collect() 
```
This is because `trunc(string, string)` isn't correctly handled. We could find a way around this and get it working, but after having a discussion with @cloud-fan, @gatorsmile, @rednaxelafx and Reynold, we decided to add `date_trunc` to be compatible with Postgres for now.

**Comment by HyukjinKwon** [comment]
> after having a discussion with @cloud-fan, @gatorsmile, @rednaxelafx and Reynold

Where did the discussion happen? Was this offline discussion? I also want to actively join in the discussion. Many implementations of the trunc works differently and I think we decide the current behaviour after sufficient discussion.

If we don't fix the stuff about #14788 in 2.3.0 timeline, it could be even more difficult because we need to keep the previous behaviour.

**Comment by HyukjinKwon** [comment]
OK. I am fine if you all guys strongly feel about this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -1295,87 +1295,184 @@ case class ParseToTimestamp(left: Expression, format: Option[Expression], child:
...
+   */
+  protected def evalHelper[T](input: InternalRow, maxLevel: Int)(
+    truncFunc: (Any, Int) => T): Any = {
     val level = if (format.foldable) {
       truncLevel
     } else {
       DateTimeUtils.parseTruncLevel(format.eval().asInstanceOf[UTF8String])
     }
-    if (level == -1) {
+    if (level == DateTimeUtils.TRUNC_INVALID || level > maxLevel) {
```
Comment (gczsjdy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: `// unknown format or too small level`?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -2797,6 +2797,21 @@ object functions {
     TruncDate(date.expr, Literal(format))
   }
 
+  /**
+   * Returns timestamp truncated to the unit specified by the format.
+   *
+   * @param format: 'year', 'yyyy', 'yy' for truncate by year,
+   *               'month', 'mon', 'mm' for truncate by month,
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: nit: one space each more.

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -1111,6 +1111,24 @@ def trunc(date, format):
     return Column(sc._jvm.functions.trunc(_to_java_column(date), format))
 
 
+@since(2.3)
+def date_trunc(format, timestamp):
+    """
+    Returns timestamp truncated to the unit specified by the format.
+
+    :param format: 'year', 'YYYY', 'yy', 'month', 'mon', 'mm',
+        'DAY', 'DD', 'HOUR', 'MINUTE', 'SECOND', 'WEEK', 'QUARTER'
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Could we make those lowercased too?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 4]
```diff
@@ -2797,6 +2797,21 @@ object functions {
...
+   *               'day', 'dd' for truncate by day,
+   *               Other options are: second, minute, hour, week, month, quarter
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Maybe, `'second', 'minute', 'hour', 'week', 'month' and 'quarter'`

**Review by HyukjinKwon — COMMENTED** [review]
Just took a quick pass.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 5]
```diff
@@ -1295,87 +1295,184 @@ case class ParseToTimestamp(left: Expression, format: Option[Expression], child:
...
         null
       } else {
-        DateTimeUtils.truncDate(d.asInstanceOf[Int], level)
+        truncFunc(d, level)
       }
     }
   }
 
-  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+  protected def codeGenHelper[T](
```
Comment (gczsjdy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Why do we need a type parameter `T`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 6]
```diff
@@ -944,9 +954,16 @@ object DateTimeUtils {
...
+  // Visible for testing.
+  val TRUNC_TO_YEAR = 1
+  val TRUNC_TO_MONTH = 2
+  val TRUNC_TO_DAY = 3
+  val TRUNC_TO_HOUR = 4
+  val TRUNC_TO_MINUTE = 5
+  val TRUNC_TO_SECOND = 6
+  val TRUNC_TO_WEEK = 7
+  val TRUNC_TO_QUARTER = 8
+  val TRUNC_INVALID = -1
```
Comment (gczsjdy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Can we bring quarter and week forward, maybe to 3 and 4? Then it's more conform to the order of time granularity and max-level design is not influenced.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 7]
```diff
@@ -1295,87 +1295,184 @@ case class ParseToTimestamp(left: Expression, format: Option[Expression], child:
...
+  /**
+   *
+   * @param input
+   * @param maxLevel Maximum level that can be used for truncation (e.g MONTH for Date input)
+   * @param truncFunc
+   * @tparam T
+   * @return
+   */
+  protected def evalHelper[T](input: InternalRow, maxLevel: Int)(
+    truncFunc: (Any, Int) => T): Any = {
```
Comment (gczsjdy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Maybe `truncFunc: (Any, Int) => Any` is enough? So we don't need to use the `T`, but I'm not sure if this is better...

**Comment by srowen** [comment]
Yeah keep any substantive discussion on the public lists. Sometimes a side conversation happens; summarize the points here.

We've rejected a lot of other functions that other DBs, but not Hive, support. Spark mostly follows Hive, and for everything else, there are UDFs. I'm not against this so much as not clear why it's exceptional

**Comment by cloud-fan** [comment]
We had an offline discussion and wanna send this out to get more feedbacks. So generally just adding `date_trunc` is pretty straightforward and makes Spark consistent with other databases about this function, while extending `trunc` to support timestamp type is a better API design.

**Comment by HyukjinKwon** [comment]
If we haven't get a similar function, I would have gone +1 but what I am less sure is `date_trunc` actually quite sounds a better version of `trunc` to be honest. Seems both also extend the same parent here `TruncTime`.

I feel like we are trying to add this better version alone by working around because it takes a relatively larger change to update other related functions consistently.

**Comment by HyukjinKwon** [comment]
I get `date_trunc` is common in other DBMS. I can see that this can be done now and we can still proceed `trunc`, etc. later. So, I am fine but still less sure tho.

**Comment by youngbink** [comment]
hmm...even if we decide to change this later, I honestly think merging `trunc` and `date_trunc` would be a simple work that only touches a couple of files, mostly in `datetimeExpressions.scala`.
This PR isn't too small as you said, but most of the codes here can be used without modification when we merge `date_trunc`.

**Comment by HyukjinKwon** [comment]
SPARK-17174 originally described few functions related with hour, min, etc. but I received an advice to fix up other related functions too even though they could also be done alone too. I agreed with doing other functions too at that time and I tried to propose as so.

I am saying I think this PR actually more targets adding another (better) version of `trunc` to support day, hour, min, etc. in the format. In this case, I think we should deduplicate the logics with related functions too.

Ah, so, I think I am less sure about why this should be done alone leaving out other related changes, and other functions we (I) usually reject.

and I think you and @cloud-fan say the reasons are, it's common and this PR targets a separate functionality consistent with other DBMS.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 8]
```diff
@@ -1295,87 +1295,184 @@ case class ParseToTimestamp(left: Expression, format: Option[Expression], child:
...
 }
 
+/**
+ * Returns date truncated to the unit specified by the format.
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """
+    _FUNC_(date, fmt) - Returns `date` with the time portion of the day truncated to the unit specified by the format model `fmt`.
+    `fmt` should be one of ["YEAR", "YYYY", "YY", "MON", "MONTH", "MM"]
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Let us use the lower case and also update the other functions in this file. For example, `ToUnixTimestamp`

**Comment by gatorsmile** [comment]
- The API proposed by this PR is consistent with the other DBs. 
- The implementation does not introduce the behavior changes. 

The implementation is clean and the PR quality is pretty good.

File: python/pyspark/sql/functions.py [hunk 9]
```diff
@@ -1111,6 +1111,24 @@ def trunc(date, format):
     return Column(sc._jvm.functions.trunc(_to_java_column(date), format))
 
 
+@since(2.3)
+def date_trunc(format, timestamp):
+    """
+    Returns timestamp truncated to the unit specified by the format.
+
+    :param format: 'year', 'YYYY', 'yy', 'month', 'mon', 'mm',
```
Comment (gatorsmile) [inline python/pyspark/sql/functions.py]: Nit: `YYYY` -> `yyyy`

Comment (gatorsmile) [reply python/pyspark/sql/functions.py]: Also update the original `trunc`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 10]
```diff
@@ -1295,87 +1295,183 @@ case class ParseToTimestamp(left: Expression, format: Option[Expression], child:
...
-// scalastyle:on line.size.limit
-case class TruncDate(date: Expression, format: Expression)
-  extends BinaryExpression with ImplicitCastInputTypes {
-  override def left: Expression = date
-  override def right: Expression = format
-
-  override def inputTypes: Seq[AbstractDataType] = Seq(DateType, StringType)
-  override def dataType: DataType = DateType
+trait TruncInstant extends BinaryExpression with ImplicitCastInputTypes {
+  val time: Expression
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Maybe, `time` -> `instant`.

**Review by HyukjinKwon — COMMENTED** [review]
Change itself seems fine.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 11]
```diff
@@ -1295,87 +1295,183 @@ case class ParseToTimestamp(left: Expression, format: Option[Expression], child:
...
   private lazy val truncLevel: Int =
     DateTimeUtils.parseTruncLevel(format.eval().asInstanceOf[UTF8String])
 
-  override def eval(input: InternalRow): Any = {
+  /**
+   *
+   * @param input internalRow (time)
+   * @param maxLevel Maximum level that can be used for truncation (e.g MONTH for Date input)
+   * @param truncFunc function: (time, level) => time
+   * @return
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Remove `@return`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 12]
```diff
@@ -1295,87 +1295,183 @@ case class ParseToTimestamp(left: Expression, format: Option[Expression], child:
...
+  val format: Expression
   override def nullable: Boolean = true
-  override def prettyName: String = "trunc"
 
   private lazy val truncLevel: Int =
     DateTimeUtils.parseTruncLevel(format.eval().asInstanceOf[UTF8String])
 
-  override def eval(input: InternalRow): Any = {
+  /**
+   *
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Remove this line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 13]
```diff
@@ -1295,87 +1295,181 @@ case class ParseToTimestamp(left: Expression, format: Option[Expression], child:
...
+  }
+}
+
+/**
+ * Returns timestamp truncated to the unit specified by the format.
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """
+    _FUNC_(fmt, date) - Returns timestamp `ts` truncated to the unit specified by the format model `fmt`.
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: `date` -> `ts`.

File: python/pyspark/sql/functions.py [hunk 14]
```diff
@@ -1111,6 +1111,24 @@ def trunc(date, format):
...
 
+@since(2.3)
+def date_trunc(format, timestamp):
+    """
+    Returns timestamp truncated to the unit specified by the format.
+
+    :param format: 'year', 'yyyy', 'yy', 'month', 'mon', 'mm',
+        'day', 'dd', 'hour', 'minute', 'second', 'week', 'quarter'
+
+    >>> df = spark.createDataFrame([('1997-02-28 05:02:11',)], ['d'])
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: `d` -> `t` or `ts`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 15]
```diff
@@ -563,6 +563,76 @@ class DateTimeUtilsSuite extends SparkFunSuite {
     }
   }
 
+  test("truncTimestamp") {
+    def test(
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: `test` -> `testTrunc` ?

**Review by HyukjinKwon — COMMENTED** [review]
Only minor nits

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master

**Pull request closed without merging**